File: back-end/src/main/java/nl/hva/backend/domain/models/game/action/impl/money/moneyExchangeImpl/SkipTurnTaxAction.java
package nl.hva.backend.domain.models.game.action.impl.money.moneyExchangeImpl;

import lombok.NoArgsConstructor;
import nl.hva.backend.domain.models.game.Player;
import nl.hva.backend.domain.models.game.action.impl.money.MoneyExchangeAction;

import javax.persistence.Entity;

/**
 * Representation, of the action, of a tax on skipping a turn, by for example not throwing dice, or picking a card.
 * This tax is paid by a player to the bank.
 *
 * @author Hamza el Haouti
 */
@Entity
@NoArgsConstructor
public class SkipTurnTaxAction extends MoneyExchangeAction {

    /**
     * Constructor for the manual creation of an instance, that is intended to be persisted to JPA.
     *
     * @param payee  The party paying, a Player instance.
     * @param amount The tax amount to be paid to the bank.
     */
    public SkipTurnTaxAction(
            Player payee,
            double amount
    ) {
        super(payee, null, amount);
    }

}

File: front-end/src/app/domain/models/game/action/impl/money/SkipTurnTaxAction.ts
import {MoneyExchangeAction} from "./MoneyExchangeAction";
import {Player} from "../../../Player";

/**
 * Representation, of the action, of a tax on skipping a turn, by for example not throwing dice, or picking a card.
 * This tax is paid by a player to the bank.
 *
 * @author Hamza el Haouti
 */
export class SkipTurnTaxAction extends MoneyExchangeAction {

  constructor(
    id: number,
    payee: Player,
    performedAt: Date,
    amount: number
  ) {
    super(SkipTurnTaxAction.name, id, payee, performedAt, null, amount);
  }

  public static trueCopy(action: SkipTurnTaxAction): SkipTurnTaxAction | null {
    return new SkipTurnTaxAction(
      action.id,
      Player.trueCopy(action.actor),
      new Date(action.performedAt),
      action.amount
    );
  }

  public override toString(): string {
    return `${this.actor.user.username} payed a tax of ${this.amount} for not doing anything during his turn.`;
  }

}
